elastic/kibana
# [Response Ops] Keep task document when enabling/disabling rules

Resolves https://github.com/elastic/kibana/issues/110096

🚨🚨🚨 A migration has been added to add `enabled: true` to existing task manager documents with `status: 'claiming' | 'idle' | 'running'`. It sets `enabled: false` to tasks with `status: 'failed' | 'unrecognized'` This might warrant extra review :) 🚨🚨🚨

## Summary

This PR adds an `enabled` field to all task documents. This is a boolean indicating whether the task is currently enabled or disabled. The task claim query has been updated to only claim enabled tasks.

Task manager changes:
* `taskSchedule.schedule` and `taskSchedule.bulkSchedule` have been updated to set `enabled: true` for newly scheduled tasks (unless `enabled: false` is explicitly set when scheduling.
* new function `taskSchedule.bulkEnableDisable` to update this flag for specific task ids

Rules client changes:
* updated `disable` function to stop deleting the underlying task manager document when a rule is disabled (with one exception, left comment on the PR). Instead, it updates the `enabled` flag on the task document to be `false`.
* updated `enable` function to stop scheduling a new task when a rule is enabled (except when no task exists). Instead, it updates the `enabled` flag on the task document to be `true`.

## To Verify

1. Run an older version of Kibana and create some rules with actions. Force a failure in an action in order to get a task doc with status `failed`. Create some rules and disable at least one of them.
2. Run this PR with the same data as step1. Verify that all task docs have a new `enabled` field. Verify that failed tasks have `enabled: false` but other tasks have `enabled: true`. Verify that the rules that were enabled before are still running.
3. Enable the disabled rules from before. Verify that a task doc gets created for these.
4. Create some new rules. Verify that enabling and disabling both old and new rules does not delete the task doc, it just flips the `enabled` flag on the task.

### Checklist
